cpp/src/arrow/util/thread_pool.h

```diff
@@ -233,15 +233,7 @@ class ARROW_EXPORT SerialExecutor : public Executor {
     if (final_fut.is_finished()) {
       return final_fut.result();
     }
-    final_fut = final_fut.Then(
-        [this](const T& res) {
-          MarkFinished();
-          return res;
-        },
-        [this](const Status& st) -> Result<T> {
-          MarkFinished();
-          return st;
-        });
+    final_fut.AddCallback([this](const Result<T>&) { MarkFinished(); });
```

Comment:
There are some differences between `AddCallback` and `Then` here.  Callbacks are run after the future waiters are notified.  This means that `final_fut.result()` would return before `MarkFinished()` has run.  It "shouldn't" matter since it would probably be bad form to mark the final future complete while there are other tasks lying around on the executor.  However, I wasn't sure it would always be safe.  For example, whomever called `MarkFinished` is going to destruct themselves and it might be nice to ensure that happens before we return from `result()`.

On the other hand, by using `Then` we ensure that `MarkFinished()` has completed executing before `final_fut.result()` returns.

So I'm not certain this change is a problem but I feel it does add risk that we don't need.

Reply:
Actually, I guess we can't return from `RunLoop` until `MarkFinished` has completed so `final_fut.result()` is always going to be called on a completed future either way.  Nevermind.  This should be ok.

Reply:
Yes, that's part of my changes (see comment about locking).

Reply:
Ah, so `wait_for_tasks` could be in the process of being destroyed which is UB given we're still in the middle of a call.

Reply:
Ah, sure enough...
> Notifying while under the lock may nevertheless be necessary when precise scheduling of events is required, e.g. if the waiting thread would exit the program if the condition is satisfied, causing destruction of the notifying thread's condition_variable.

From https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one